torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -200,18 +200,38 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
           autogradContext != nullptr,
           "autogradContext is nullptr, FORWARD_AUTOGRAD_REQ should always get "
           "or create valid autogradContext in addRecvRpcBackward.");
-      autogradContainer.setCurrentContextId(autogradContext->contextId());
+      auto contextId = autogradContext->contextId();
+      autogradContainer.setCurrentContextId(contextId);
 
       // Process the original RPC.
       auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
+      auto fromWorkerId = rpcWithAutograd.fromWorkerId();
       auto wrappedRpcResponse = processRpc(
           rpcWithAutograd.wrappedRpc(), wrappedMessageType, messageId);
-      wrappedRpcResponse->waitNoThrow(); // TODO: make async
 
-      return wrap(getMessageWithAutograd(
-          rpcWithAutograd.fromWorkerId(),
-          std::move(*wrappedRpcResponse).moveValue(),
-          MessageType::FORWARD_AUTOGRAD_RESP));
+      // Create a future that handles the response.
+      auto responseFuture = std::make_shared<FutureMessage>();
+      wrappedRpcResponse->addCallback(
+          [responseFuture,
+           messageId,
+           wrappedRpcResponse,
+           fromWorkerId,
+           contextId](
+              const Message& /* unused */,
+              const c10::optional<utils::FutureError>& err) {
+            ContextPusher pusher(contextId);
+            if (err) {
+              responseFuture->setError(*err);
+            } else {
+              Message m = getMessageWithAutograd(
+                  fromWorkerId,
+                  std::move(*wrappedRpcResponse).moveValue(),
+                  MessageType::FORWARD_AUTOGRAD_RESP);
+              m.setId(messageId);
+              responseFuture->markCompleted(std::move(m));
+            }
+          });
+      return responseFuture;
```

Comment:
What is the benefit of making this async vs just waiting? `processRpc` mostly just does CPU work

Reply:
Agree that it does largely do cpu work currently, and probably won't be as impactful as I hoped when I wrote up the diff.

fwiw, my original motivation here is simply to get rid of gratuitous synchronous-ness in this handler function; I've been trying to send out PRs (some bite-sized like this) and expedite reviews of PRs that help improve parallelism without throwing hundreds of mostly-waiting threads at the problem. Since long-term, zillions-of-threads is a recipe for memory thrash.

But... given the test failures (that I thought I had working locally when running yesterday, not sure what's going on), this might be lower priority, and might shelve this unless I can untangle a bit.